code/modules/events/crystal_event.dm

```diff
@@ -0,0 +1,477 @@
+					to_chat(user, "<span class='notice'>The [src] has been restored and restabilized!</span>")
+					playsound(get_turf(src), 'sound/effects/supermatter.ogg', 35, TRUE)
+					injector.filled = FALSE
+					active = FALSE
+					restore()
+
+///Restore the Destabilized Crystal as it was before
+/obj/machinery/destabilized_crystal/proc/restore()
+	priority_announce("The Crystal has been restored and is now stable again, your sector of space is now safe from the ZK-Lambda-Class Scenario, \
+						kill the remaining crystal monsters and go back to work")
+	sound_to_playing_players('sound/misc/notice2.ogg')
+	var/turf/T = get_turf(src)
+	new/obj/machinery/power/supermatter_crystal(T)
+	for(var/Portal in GLOB.crystal_portals)
+		qdel(Portal)
+	qdel(src)
+
+/*
+This section is for the crystal stabilizer item
+*/
+/obj/item/crystal_stabilizer
+	name = "Supermatter Stabilizer"
+	desc = "Used when the Supermatter Matrix is starting to reach the destruction point."
+	icon = 'icons/obj/supermatter.dmi'
+	icon_state = "stabilizer"
+	slot_flags = ITEM_SLOT_BACK
+	w_class = WEIGHT_CLASS_HUGE
+	///The stabilizer is one use only
+	var/filled = FALSE
+
+/obj/item/crystal_stabilizer/attackby(obj/item/W, mob/living/user, params)
+	if(!istype(W) || (W.item_flags & ABSTRACT) || !istype(user))
+		return
+	if(istype(W, /obj/item/assembly/signaler/anomaly))
+		if(filled)
+			return
+		to_chat(user, "<span class='notice'>You refill the [src]</span>")
+		filled = TRUE
+		qdel(W)
+
+/*
+This section is for the crystal portals variations
+*/
+/obj/structure/crystal_portal
+	name = "crystal portal"
+	desc = "this shouldn't be here"
+	icon = 'icons/effects/effects.dmi'
+	icon_state = "anom"
+	color = "#B2FFFE"
+	anchored = TRUE
+	resistance_flags = INDESTRUCTIBLE | LAVA_PROOF | FIRE_PROOF | UNACIDABLE | ACID_PROOF | FREEZE_PROOF
+	///Max amount of mobs that a portal can spawn in any given time
+	var/max_mobs = 5
+	///Spawn time between each mobs
+	var/spawn_time = 1000
+	///Type of mob that the portal will spawn, if more than one type in the list will choose randomly
+	var/mob_types = list(/mob/living/simple_animal/hostile/carp)
+	///Fluff text for each mob spawned
+	var/spawn_text = "emerges from"
+	///Affiliation to a faction, used to stop mobs from destroying their portals (unused for now)
+	var/faction = list("hostile")
+	///Spawner component
+	var/spawner_type = /datum/component/spawner
+	///This var check if the portal has been closed by a player with a neutralizer
+	var/closed = FALSE
+
+/obj/structure/crystal_portal/Initialize()
+	. = ..()
+	AddComponent(spawner_type, mob_types, spawn_time, faction, spawn_text, max_mobs)
+	GLOB.crystal_portals += src
+
+/obj/structure/crystal_portal/Destroy()
+	GLOB.crystal_portals -= src
+	if(!closed)
+		switch(name)
+			if("Small Portal")
+				explosion(loc, 0,1,3)
+			if("Medium Portal")
+				explosion(loc, 0,3,5)
+			if("Big Portal")
+				explosion(loc, 1,3,5)
+			if("Huge Portal")
+				explosion(loc, 2,5,7)
```

Comment:
Fixing the crystal before closing all the portals will cause them to explode. Is that intentional?

Reply:
yes

Reply:
This will happen before the crystal portal is moved to nullspace, meaning the explosion will affect it.
Avoid adding these interactions with the outside world on `Destroy()` if you can avoid it, instead make a precursor proc that calls for its deletion.